activerecord/lib/active_record/associations/builder/belongs_to.rb

```diff
@@ -110,5 +110,24 @@ def self.add_destroy_callbacks(model, reflection)
       name = reflection.name
       model.after_destroy lambda { |o| o.association(name).handle_dependency }
     end
+
+    def self.define_validations(model, reflection)
+      if reflection.options.key?(:required)
+        ActiveSupport::Deprecation.warn('`required` option for belongs_to relation is deprecated and replaced with `optional` option.')
```

Comment:
'`required: true` is now the default for belongs_to and can be removed.' – don't even need to talk about optional here. What they wanted to do is now being done for them.